### PR TITLE
fix: enum self reference

### LINKED ISSUE #47285: Enum internal reference points to the wrong value.
# Bug Report

### 🔎 Search Terms
- Enum 

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

#### I tried the latest version 4.5.4 and version 3.3.3 as well as nightly 4.6.0-dev.20211231 in Playground.
All successfully reproduced this issue.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/KYOwrgtgBAglDeAoKKoEMoF4oCYA0yqARllAMwGpQAepaAdEZagJ6lH1rMoBepL9WgCp0gggF8gA)

### 💻 Code

```ts
enum A {
    a = 2,
    b = 3,
    x = a.b,
    y = b.a,
    z = y.x * a.x,
}
```

### 🙁 Actual behavior

```js
var A;
(function (A) {
    A[A["a"] = 2] = "a";
    A[A["b"] = 3] = "b";
    A[A["x"] = 3] = "x";
    A[A["y"] = 2] = "y";
    A[A["z"] = 9] = "z";
})(A || (A = {}));
```

### 🙂 Expected behavior

- Keep `a.b`, `b.a`, `y.x * a.x` unchanged OR they should be qualified to `EnumName.MemberName` like `A.a.b`,`A.b.a`,`A.y.x`, `A.a.x`
-  report an error.

### PR DESCRIPTION
**Pull request by magic-akari**
- Fixes #47285

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -39432,11 +39431,33 @@ namespace ts {
...
+                const expr = node as ElementAccessExpression;
+
+                if (expr.expression.kind === SyntaxKind.Identifier && isStringLiteralLike(expr.argumentExpression)) {
+                    return true;
+                }
+
+                return isConstantMemberAccess(expr.expression);
+            }
+
+            return false;
```
Comment (sandersn) [inline src/compiler/checker.ts:39460]: on reading this, I *think* the only change is to ensure that returning true always means that `node: AccessExpression`. If so, I don't think it's worth the refactor. I'd prefer to keep the code the way it was and just check that `node` has a valid type.

**Review by sandersn — CHANGES_REQUESTED** [review]

File: tests/baselines/reference/constEnumErrors.errors.txt [hunk 2]
```diff
@@ -37,9 +39,13 @@ tests/cases/compiler/constEnumErrors.ts(43,9): error TS2478: 'const' enum member
         Y = E1.Z,
             ~~~~
 !!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
+               ~
```
[the remark this answers is not in the document]
Comment (magic-akari) [reply tests/baselines/reference/constEnumErrors.errors.txt:42]: I have double-checked. This is an issue that should be reported. 
It is also an issue that falls under the scope of this fix.
Z doesn't exist, so should we ignore it?

File: src/compiler/checker.ts (hunk 1 above)
Comment (magic-akari) [reply src/compiler/checker.ts:39460]: Do you mean that `ElementAccessExpression | PropertyAccessExpression` and `AccessExpression` are equivalent?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -39432,11 +39431,33 @@ namespace ts {
...
         }
 
-        function isConstantMemberAccess(node: Expression): boolean {
-            return node.kind === SyntaxKind.Identifier ||
-                node.kind === SyntaxKind.PropertyAccessExpression && isConstantMemberAccess((node as PropertyAccessExpression).expression) ||
-                node.kind === SyntaxKind.ElementAccessExpression && isConstantMemberAccess((node as ElementAccessExpression).expression) &&
-                    isStringLiteralLike((node as ElementAccessExpression).argumentExpression);
+        function isConstantMemberAccess(node: Expression): node is ElementAccessExpression | PropertyAccessExpression {
+            const type = getTypeOfExpression(node);
+            if(!type.symbol) {
```
[the remark this answers is not in the document]
Comment (magic-akari) [reply src/compiler/checker.ts:39436]: When the type of a node is inferred to be `any`, there is no symbol.
Is there a more direct way to determine if the type is `any`, other than by detect symbol?
If so, please let me know.

File: src/compiler/checker.ts (hunk 1 above)
Comment (magic-akari) [reply src/compiler/checker.ts:39460]: No, the original code returns true directly when the node is `SyntaxKind.Identifier`, which needs to be fixed.

File: tests/baselines/reference/constEnumErrors.errors.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sandersn) [reply tests/baselines/reference/constEnumErrors.errors.txt:42]: No, you're right, it makes sense to report both in this case.

File: src/compiler/checker.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandersn) [reply src/compiler/checker.ts:39436]: I think you could check `type === errorType` &mdash; `errorType` is the internal "error any" type.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts:39460]: Can you give an example of an input that's correct now and incorrect on the old code?

`isConstantMemberAccess` is initially called only with an `AccessExpression` so the `Identifier` check will never initially be true.

**Review by sandersn — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (magic-akari) [reply src/compiler/checker.ts]: You are right.
After I revert the `isConstantMemberAccess` logic change, all the test cases pass.
It seems that the issue is only caused by `errorType`.

**Review by sandersn — APPROVED** [review]

**Pull request merged**